pkgs/development/cuda-modules/ON-CROSS-COMPILATION.md

```diff
@@ -0,0 +1,94 @@
+# On cross-compilation
+
+## Summary
+
+The cross-compilation work for the Nixpkgs CUDA ecosystem has hit a blocker --
+one which has no clear resolution.
+
+The CUDA redistributables[^1], the individually packaged pieces of software
+historically only made available through the CUDA Toolkit, are not equipped for
+cross-compilation. Cross-compilation appears to be possible only with special
+versions of the Debian CUDA Toolkit installer[^2], rather than the CUDA
+redistributables. Given the NixOS CUDA Team's focus on moving to the CUDA
+redistributables and eventually dropping support for the CUDA Toolkit runfile
+installer, this is a setback for the future of cross-compilation of
+CUDA-accelerated packages in Nixpkgs.
+
+## Details
+
+### Background
+
+Cross-compilation holds varying levels of importance for the diverse user base
+of Nixpkgs. It allows for the building of software for a different architecture
+than the one on which the build is taking place. This is particularly useful for
+embedded systems, where the build machine may be a powerful x86_64 machine, but
+the target is an ARM device. Unfortunately, such capabilities come at a cost:
+cross-compilation is a complex and error-prone process, and while Nixpkgs may
+make it possible, it certainly does not make it easy.
+
+With the 11.4 release of the CUDA Toolkit, NVIDIA introduced redistributables
+that allow for the installation of individual components of the CUDA Toolkit
+without the need for the monolithic installer. Since their introduction, the
+NixOS CUDA Team has made it a priority to migrate to these smaller, more
+manageable packages. The redistributables provide a number of benefits over the
+monolithic installer, including a staggering reduction in closure sizes and
+fine-grained control over dependencies. Until the realization that
+cross-compilation is not possible with the redistributables, there were no clear
+benefits, only drawbacks, to using or maintaining the CUDA Toolkit runfile
+installer in Nixpkgs.
+
+### Problem
+
+As the only way to cross-compile CUDA software, further work on
+cross-compilation would necessitate packaging the Debian CUDA Toolkit installers
+-- a highly non-trivial task. Beyond the difficulties inherent in repackaging a
```

Comment:
Is that a fact? Does compilation fail if you take the nvcc binary from the build platform and the libraries from the host platform?

Reply:
Yes, it does :(

The cross-compilation CUDA Toolkit Debian installers have additional directories in the root for different hosts and targets.

The redistributables don’t have anything like that and its NVCC dies when encountering libraries built for a different arch.

Reply:
Superficially, I can say there's something going on with `host-*` and `target-*` subdirectories present in the CUDA Toolkit (of all kinds, I believe, cross or not).

In the non-cross CUDA Toolkit, on `x86_64-linux` (that's the NixOS system name) there will be a `host-linux-x64` and `target-linux-x64` subdirectory in the root of the CUDA Toolkit.

In the cross compilation CUDA Toolkit, the values might well be something like `host-linux-x64` with `target-linux-sbsa` (ARM servers) or `target-linux-aarch64` (Jetson devices).

I'd have to spin up an Ubuntu system to install the cross-compilation version of the CUDA Toolkit and trace through the flow of CMake to see how it's driving cross-compilation.

I fear that beyond the presence of these host and target directories, the binaries in the cross-compilation version of the CUDA Toolkit are different from those offered in the redistributables -- though I haven't done any sort of analysis on the individual binaries themselves, I worry they're compiled differently from NVIDIA's sources in such a way that we cannot mimic the changes necessary for cross-compilation in an ad-hoc manner.

At any rate, it would require more time than I'm able to allocate.

Reply:
> the binaries in the cross-compilation version of the CUDA Toolkit are different from those offered in the redistributables

We can also literally look at the diffoscope to see if that's the case (not a legal advise xDDD)

Reply:
I'll add additional info to that doc to hint at next steps or possible things to try ;)

Reply:
> add

https://www.nature.com/articles/s41586-021-03380-y 🙃

Reply:
Hmmm, I'm not entirely sure why these issues are cropping up, seems to be related to the `targets/` directory symlinks.

The binaries are the same, they are built once per platform/version and then packaged into each format; some formats (Conda and Python wheels) involve `patchelf`.

I would expect the "binary archive" tarballs, Debian packages, and RPM packages to be bit-identical.

The cross-packages are more-or-less a subset of the native packages; some libraries provide stub files, while others use the native runtime.

Reply:
Thanks for this feedback @kmittman! How would you recommend best approaching cross compilation with the redistributable packages? Shall we copy `targets/` things from the CUDA toolkit package then?

Reply:
We really should (have) start(ed) with an MWE: piecing together an nvcc derivation sufficient to run `nvcc` directly on a single file (saxpy but without cmake)

Reply:
Moved to #303122